migrations/back-37/back-37.go

```diff
@@ -0,0 +1,207 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/urfave/cli"
+	"go.mongodb.org/mongo-driver/bson"
+
+	"github.com/tidepool-org/platform/application"
+	"github.com/tidepool-org/platform/data/blood/glucose"
+	"github.com/tidepool-org/platform/errors"
+	migrationMongo "github.com/tidepool-org/platform/migration/mongo"
+	storeStructuredMongo "github.com/tidepool-org/platform/store/structured/mongo"
+)
+
+func main() {
+	application.RunAndExit(NewMigration())
+}
+
+type Migration struct {
+	*migrationMongo.Migration
+	dataRepository *storeStructuredMongo.Repository
+}
+
+func NewMigration() *Migration {
+	return &Migration{
+		Migration: migrationMongo.NewMigration(),
+	}
+}
+
+func (m *Migration) Initialize(provider application.Provider) error {
+	if err := m.Migration.Initialize(provider); err != nil {
+		return err
+	}
+
+	m.CLI().Usage = "BACK-37: Migrate all existing data to add required Platform deduplication hash fields"
+	m.CLI().Description = "BACK-37: To fully migrate devices from the `jellyfish` upload API to the `platform` upload API"
+	m.CLI().Authors = []cli.Author{
+		{
+			Name:  "J H BATE",
+			Email: "jamie@tidepool.org",
+		},
+	}
+
+	m.CLI().Action = func(ctx *cli.Context) error {
+		if !m.ParseContext(ctx) {
+			return nil
+		}
+		return m.execute()
+	}
+
+	return nil
+}
+
+func (m *Migration) execute() error {
+	m.Logger().Debug("Migrate jellyfish upload API data")
+	m.Logger().Debug("Creating data store")
+
+	mongoConfig := m.NewMongoConfig()
+	mongoConfig.Database = "data"
+	mongoConfig.Timeout = 60 * time.Minute
+	dataStore, err := storeStructuredMongo.NewStore(mongoConfig)
+	if err != nil {
+		return errors.Wrap(err, "unable to create data store")
+	}
+	defer dataStore.Terminate(context.Background())
+
+	m.Logger().Debug("Creating data repository")
+
+	m.dataRepository = dataStore.GetRepository("deviceData")
+	hashUpdatedCount, archivedCount, errorCount := m.migrateJellyfishDocuments()
+	m.Logger().Infof("Migrated %d jellyfish documents", hashUpdatedCount)
+	m.Logger().Infof("Archived %d duplicate jellyfish documents", archivedCount)
+	m.Logger().Infof("%d errors occurred", errorCount)
+
+	return nil
+}
+
+func (m *Migration) migrateJellyfishDocuments() (int, int, int) {
+	logger := m.Logger()
+
+	logger.Debug("Finding distinct users")
+
+	var hashUpdatedCount, archivedCount, errorCount int
+
+	userIDs, err := m.dataRepository.Distinct(context.Background(), "_userId", bson.M{})
+	if err != nil {
+		logger.WithError(err).Error("Unable to execute distinct query")
+	} else {
+		logger.Debugf("Finding jellyfish records for %d users", len(userIDs))
+
+		for _, userID := range userIDs {
+			logger.Debugf("Finding jellyfish records for user ID %s", userID)
+			selector := bson.M{
+				"_userId": userID,
+				"_active": true,
+				// uploads aren't de-duped.
+				"type":          bson.M{"$ne": "upload"},
```

Comment:
Are `upload` records in the data collection now? I thought they were moved.

Reply:
Or are they still around, but just not used in the data collection?

Reply:
they have moved